libcxx/test/std/utilities/meta/meta.trans/visibility.sh.cpp

```diff
@@ -0,0 +1,71 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17
+
+// Make sure that the types and variables have the correct visibility attributes
+
+// RUN: %{cxx} %s %{flags} %{compile_flags} %{link_flags} -DSHARED -fPIC -fvisibility=hidden -shared -o %t.shared_lib
+// RUN: %{build} -fvisibility=hidden %t.shared_lib
+// RUN: %{run}
+
+#include <algorithm>
+#include <cassert>
+#include <type_traits>
+#include <vector>
+
+[[gnu::visibility("default")]] extern std::vector<const std::type_info*> shared_lib_type_infos;
+
+struct [[gnu::visibility("default")]] S {
+  static constexpr bool value = false;
+};
+
+inline std::vector<const std::type_info*> get_type_infos() {
+  return {
+      &typeid(std::remove_const<int>),
```

Comment:
I really question the value of this. For the inline variables, sure, but do we really want to get into the business of providing this guarantee for arbitrary type traits? The value this provides to users seems to be really small.

By the way, do we have the same problem for nearly all other types in the library?

Reply:
Technically we have to do that to be conforming I think.

Depends on what you mean by problem. We have default visibility for all the type symbols, at least on clang, so this is really just testing what we do already here. If you mean that we don't test this for any type in the library, then yes, this is the first test to make sure any type symbols have default visibility.

Reply:
This has the potential to get really out of hands for a questionable benefit, don't you think? Some types definitely need their visibility to be correct, like exception types. If we make a mistake on one of those, the exception handling mechanism won't work properly. But for e.g. type traits, this really doesn't matter that much. I think that's why I'm surprised you're starting with those tests.

If we follow the reasoning here, we would end up adding a test that every type we add to the library has the right visibility. And we would add a new test like that whenever we implement a new user-visible type. Is that what you have in mind?

Reply:
I agree that type traits aren't the most useful to have correct RTTI for. I really just started with those because I had to write a test for the variables anyways. Adding a test for the classes was not significantly more work than copy-pasting the code. OTOH I could see a use-case for pretty much every non-empty type, so adding the few empty types doesn't seem like that much extra work for a tick-mark in "fully conforming across dylib boundaries".

I think it would be good to add tests for the important types (`string`, `vector`, exception types, etc.), since I did almost regress on this a few times in the past. Then again, the situation has changed now, since all types are properly annotated through `_LIBCPP_BEGIN_NAMESPACE_STD` except when using GCC. So maybe it would be good enough to add a clang-tidy test ensuring all public types have the proper annotations.

FYI we'll probably have a longer discussion on GCC and ABI guarantees tomorrow, so maybe we could talk about this more at that point if there is some time.

Reply:
Looking forward to that discussion. Those are always a treat!

Reply:
Per live review, you are correct that this is needed for standards conformance. However, the question of whether we care enough about this aspect of conformance to take (a hypothetical) load-time performance hit is something worth asking.

I wonder if this will truly cause an increase in the number of weak symbols in normal programs. It's OK if a program that clearly ODR-uses `foo_v` ends up with a weak-def, since that's necessary for correctness. What's bad is if idiomatic usage of these traits end up generating weak-defs that can't be removed: then we'd introduce a load-time regression for the purpose of standards conformance, but for an aspect of conformance that provides very very marginal actual value.

In order to know whether this is worth it, could you please check in which cases a weak def is actually emitted? Also, I think it could be insightful to run the test suite and `nm` the executables produced while looking for weak-defs. If the number of weak-defs increases significantly after this change, that's some data point we could use to make a decision. If it doesn't, and if Clang seems to not be emitting weak-defs in "useless" cases, then I don't think there's much to push back against this change.

The level of testing is another aspect we could discuss if we decide to move forward, but I think that's a minor point.

-----------------------

For the `nm` experiment, I would personally modify `format.py` to automatically run `nm %t.exe` with some post-processing after every `.pass.cpp` test. I think that would be good enough.